### ISSUE TITLE
#[derive(Clone/Copy/etc.)] is confusing

### ISSUE THREAD

**Issue by yshui** [rust-lang/rust#49787]
The wording in the [document of Clone](https://doc.rust-lang.org/1.11.0/core/clone/trait.Clone.html#derivable)

> This trait can be used with #[derive] if all fields are Clone.

suggests that a struct **can't** use `#[derive(Clone)]` if some of its fields are not Clone. Which seems to imply that some kind of error will be generated if `#[derive(Clone)]` is used in that case.

In other words, someone who read the documentation would expect

```rust
#[derive(Clone)]
struct SomeStruct<T> {
    ...
}
```

to fail, Because T is not Clone.

While in reality, a `impl<T: Clone> Clone for SomeStruct<T>` is generated, so it compiles. This can sometimes be frustrating because I sometimes forget to write `T: Clone`, only to find out much later when I actually need `.clone()`.

Apparently it is impractical to change the behaviour of `#[derive]` now. But I think the document can be worded better. And perhaps an alternate version of `#[derive]` that fails early can be added.

**Comment by yshui** [comment]
This also causes unhelpful error messages.

```rust
#[derive(Debug)]
struct Test<T> {
    a: T,
}
struct Undebugable;
fn main() {
    let a = Undebugable;
    let c = Test { a: a }; // <- passes even if T is not Clone
    println!("{:?}", c);
}
```

The compiler tells me to

> add `#[derive(Debug)]` or manually implement `std::fmt::Debug`

While `#[derive(Debug)]` is already there.

**Comment by steffengy** [comment]
> The compiler tells me to add #[derive(Debug)] or manually implement std::fmt::Debug

The error message is `Undebugable doesn't implement `std::fmt::Debug`, in my opinion that's pretty clear?

**Comment by yshui** [comment]
@steffengy You are right, this is a poor example. But the original point still stands

**Comment by durka** [comment]
Can you suggest better wording for the docs to make it clear what happens?

**Comment by yshui** [comment]
@durka I think it should include that for generic case, the derivation is bound to the type parameter.

**Comment by czipperz** [comment]
This was closed by bbc8c932fb05dfaed7b5bfe252712bede722b764
